Reduction class for single-crystal diffraction data from SXD and WISH instruments

RichardWaiteSTFC opened the pull request:
**Description of work.**

Adds a base single-crystal reduction class (`BaseSX`) and an inherited class for instrument specific operation for WISH (`WishSX` and SXD (`SXD`). 

In terms of tests:

- Static functions in the base class (`BaseSX` and `SXD`) have been unit tested
- SXD systems test have been re-written to use the new class and significantly added to (this tests all the non static functions and the whole workflow)
- WISH systems test for the vanadium processing now uses the `WishSX` class

These classes are a work in progress, in particular the WISH one  - but it's important to merge this into the code base so that it can be tested by scientists during the cycle, and to unite lots of helper functions that lived in different scripts. Also to stop the two workflows (SXD and WISH evolving separately such that e.g. bug fixes being made to one script but not the other etc.)

**To test:**

(1) Run a detector calibration

```
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from Diffraction.single_crystal.base_sx import PEAK_TYPE, INTEGRATION_TYPE
from Diffraction.single_crystal.sxd import SXD

sxd = SXD()
wsname = sxd.load_run(23767)
ClearInstrumentParameters(Workspace=wsname)
pk_wsname = sxd.find_sx_peaks(wsname,  nstd=8)
sxd.remove_peaks_on_detector_edge(pk_wsname, 2)

# inital UB
FindUBUsingLatticeParameters(PeaksWorkspace=pk_wsname , a=5.6402, b=5.6402, c=5.6402, 
                             alpha=90, beta=90, gamma=90, FixParameters=False)
IndexPeaks(PeaksWorkspace=pk_wsname, Tolerance=0.15, CommonUBForAll=True)
# force cubic (does better job than FindUBUsingLatticeParameters with FixParameters=True)
CalculateUMatrix(PeaksWorkspace=pk_wsname , a=5.6402, b=5.6402, c=5.6402, 
                 alpha=90, beta=90, gamma=90)
IndexPeaks(PeaksWorkspace=pk_wsname, Tolerance=0.15, CommonUBForAll=False) # 229/290

save_dir = r"C:\Users\xhg73778\Desktop"
detcal_path = sxd.calibrate_sxd_panels(wsname, pk_wsname, save_dir, tol=0.25)

IndexPeaks(PeaksWorkspace=pk_wsname, Tolerance=0.15, CommonUBForAll=True)  # 262/290
```


(2) Run an NaCl reduction (using the `detcal_path` you made in step 1)

```
save_dir = r"C:\Users\xhg73778\Desktop"
detcal_path = r"C:\Users\xhg73778\Desktop\SXD23767_peaks_detcal.xml"
# integration args
intPeaksMDArgs = {'ellipsoid': False}
# setup class                    
sxd = SXD(vanadium_runno=32671, empty_runno=32673, detcal_path=detcal_path)
sxd.set_sample(Geometry={'Shape': 'CSG', 'Value': sxd.sphere_shape}, 
               Material={'ChemicalFormula': 'Na Cl', 'SampleNumberDensity': 0.0223})
sxd.set_goniometer_axes([0,1,0,1])  # ccw rotation around vertical
sxd.process_vanadium()

# process a single run
for run in range(33335,33336):
    sxd.process_data([run], [0]) # 0 gonio rotation
    sxd.convert_to_MD(run=run)  # convert to QLab (default)
    # find peaks
    ws = sxd.get_ws(run)
    peaks = sxd.find_sx_peaks(ws,  nstd=12)
    sxd.remove_peaks_on_detector_edge(peaks, 2)
    sxd.remove_duplicate_peaks_by_qlab(peaks, q_tol=0.05)
    sxd.set_peaks(run, peaks)
    # Find a UB (or load one etc. - will rotate UB by goniometer)
    FindUBUsingFFT(PeaksWorkspace=peaks, MinD=1, MaxD=10)
    # calibrate sample position
    sxd.calibrate_sample_pos(tol=0.15)
    # predict peaks
    sxd.predict_peaks(MinDSpacing=0.5, WavelengthMin=0.4, ReflectionCondition='All-face centred')
    # integrate peaks
    sxd.integrate_data(INTEGRATION_TYPE.MD_OPTIMAL_RADIUS, PEAK_TYPE.FOUND, 
                                  run=run, scale=12, **intPeaksMDArgs)
   sxd.save_peak_table(run, PEAK_TYPE.FOUND, INTEGRATION_TYPE.MD_OPTIMAL_RADIUS, save_dir, save_format='Jana')
```
   
(3) Check systems tests pass

Fixes #34217
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

RichardWaiteSTFC committed:
Add wish single-crystal reduction class

RichardWaiteSTFC committed:
Merge branch 'main' into reduction_class_WISH

RichardWaiteSTFC committed:
Add base single crystal class for WISh and SXD

RichardWaiteSTFC committed:
Add reduction class for SXD

RichardWaiteSTFC committed:
Use peak ws getters/setters in base class methods

RichardWaiteSTFC committed:
BUg fixes and override peak finding method in sxd class

RichardWaiteSTFC committed:
Decorator to apply fucntions to all runs by default

User passes a specific run as kwarg to override this behaviour

RichardWaiteSTFC committed:
Refactor save peaks to have separate method for combining all runs

RichardWaiteSTFC committed:
Fix bugs related to apply all

RichardWaiteSTFC committed:
Overide radius calc. in sxd class (made non-static)

RichardWaiteSTFC committed:
Add a setter for the workspace

RichardWaiteSTFC committed:
Add helper funcs to remove unindexed and non-integrated pks

RichardWaiteSTFC committed:
Disable logging for select algorithms

RichardWaiteSTFC committed:
Pass multiple gonio angle lists (or motor names) using *args

RichardWaiteSTFC committed:
USe abstract methods in BaseSX and make WishSX inherit from BaseSX

RichardWaiteSTFC committed:
Overide find peaks in WishSX

RichardWaiteSTFC committed:
Add ability to mask a number edges in WishSX

Apply mask to tubes before converting to MD and when finding peaks

RichardWaiteSTFC committed:
Add unit tests for static methods of base_sx

RichardWaiteSTFC committed:
Add unit tests for SXD class

Test also methods of BaseSX that require class instantiated (BaseSX has abstract methods and therefore cannot be instantiated)

RichardWaiteSTFC committed:
Use SXD reduction class in sxd systems test

Fixed bug where corrected for empty and vanadium twice

RichardWaiteSTFC committed:
Merge branch 'main' into reduction_class_WISH

RichardWaiteSTFC committed:
Fix SXD systems test

RichardWaiteSTFC committed:
Use reduction class in WISH vanadium processing system test

RichardWaiteSTFC committed:
Add better reporting for sxd panel calibration

DannyHindson commented on scripts/Diffraction/single_crystal/sxd.py:
```diff
@@ -0,0 +1,320 @@
+    @staticmethod
+    def calibrate_sxd_panels(ws, peaks, save_dir, tol=0.15, maxShiftInMeters=0.025):
+        """
+        Calibrate SXD panels and apply the calibration to th
```
Given this is a fairly chunky function it's worth filling out the description here

DannyHindson commented on scripts/Diffraction/single_crystal/base_sx.py:
```diff
@@ -0,0 +1,574 @@
+
+    # --- getters ---
+
+    def get_ws(self, run):
```
Does this return a ws name? Might be better to call it `get_ws_name` if yes. Similar comment for `get_md`

RichardWaiteSTFC replied:
> Does this return a ws name? Might be better to call it `get_ws_name` if yes. Similar comment for `get_md`

I would like to encourage users to use strings in their scripts (less chance of bugs through an object becoming invalidated, and easier/more transparent when assigning the output of an algorithm to have the same name) so I force the class setters to set the internal dict values with workspace names. That said the user can access the internal dict (a downside of python...) and set the workspace to a python object if they wanted - and it's designed so everything will (should!) work anyway! i.e. I assume the return from a getter or function input could be a string or python object  - hence the `BaseSX.retrieve` method is called a lot. This makes it a bit more cumbersome for the developer but I think should be more robust...

So I would say it is a getter of the dict field `ws` - but I wouldn't want to bake in an assumption as to the data type - but maybe that's just confusing? Or I should have both?

RichardWaiteSTFC replied:
Yep I've thought about this a bit more and definitely need to be clear what data type the getters return - I will force it to be a string and call put `name` in the function name. Not sure I will add getters for ws objects as well, will rely on BaseSX.retrieve if users want that for now (hopefully they won't need it now there's the helper functions)

DannyHindson commented on scripts/Diffraction/single_crystal/sxd.py:
```diff
@@ -0,0 +1,320 @@
+            nindexed, *_ = mantid.IndexPeaks(PeaksWorkspace=peaks_bank, Tolerance=tol_to_report, CommonUBForAll=False, EnableLogging=False)
+            logger.notice(f"Indexed before optimization = {nindexed}/{npeaks_bank} peaks within {tol_to_report}")
+            mantid.IndexPeaks(PeaksWorkspace=peaks_name, Tolerance=tol, CommonUBForAll=False, EnableLogging=False)
+            peaks_bank_mod, pos_table, *_ = mantid.OptimizeCrystalPlacement(
```
After running the test SXD script I see a couple of workspaces left behind called CovarianceInfo or FitInfoTable which seem to be from OptimizeCrystalPlacement. If these are from the last bank processed they perhaps don't have any special significance. So could these be deleted?

DannyHindson commented on scripts/Diffraction/single_crystal/base_sx.py:
```diff
@@ -0,0 +1,574 @@
+            md_name = wsname + "_MD"
+        xunit = BaseSX.get_xunit(wsname)  # get initial xunit
+        BaseSX._normalise_by_bin_width_in_k(wsname)  # normalise by bin-width in K = 2pi/lambda
+        wsMD = mantid.ConvertToDiffractionMDWorkspace(
```
Does this call leave a PreprocessedDetectorsWS behind? Might be worth cleaning it up at the end of the function

DannyHindson commented on scripts/Diffraction/wish/wishSX.py:
```diff
@@ -0,0 +1,129 @@
+    def __init__(self, vanadium_runno=None):
+        super().__init__(vanadium_runno)
+        self.grp_ws = None  # banks grouping workspace
+        self.ngrp = None  # no. groups (i.e. number of banks)
```
are `self.grp_ws` and `self.ngrp` ever used in this class?

DannyHindson commented on scripts/Diffraction/single_crystal/sxd.py:
```diff
@@ -0,0 +1,320 @@
+            InputWorkspace=ws,
+            OutputWorkspace="grouped",
+            IgnoreGroupNumber=False,
+            CopyGroupingFromWorkspace=self.grp_ws,
```
Perhaps the grp_ws could just be passed into this function as a parameter. Doesn't seem like it's used anywhere else so having as a field on the class may be unnecessary

DannyHindson commented on scripts/Diffraction/wish/wishSX.py:
```diff
@@ -0,0 +1,129 @@
+            # set sample (must be done after gonio to rotate shape) and correct for attenuation
+            if self.sample_dict is not None:
+                mantid.SetSample(wsname, EnableLogging=False, **self.sample_dict)
+                mantid.ConvertUnits(InputWorkspace=wsname, OutputWorkspace=wsname, Target="Wavelength", EnableLogging=False)
```
Similar comment to the one in the sxd.py unit - are the units on the ws at the end of this function different depending on whether the absorption correction is done. Think the final units are wavelength if code comes in here

DannyHindson commented on scripts/Diffraction/wish/wishSX.py:
```diff
@@ -0,0 +1,129 @@
+        ws, mon = mantid.LoadRaw(Filename=wsname + ".raw", OutputWorkspace=wsname, LoadMonitors="Separate")
+        mon_name = mon.name()
+        mantid.CropWorkspace(InputWorkspace=wsname, OutputWorkspace=wsname, XMin=6000, XMax=99000)
+        mantid.CropWorkspace(InputWorkspace=mon_name, OutputWorkspace=mon_name, XMin=6000, XMax=99000)
```
Similar comment to sxd.py. I think you should pull out these crop ranges as constants

DannyHindson reviewed (dismissed):
I've had an initial look around. It's obviously been a lot of work pulling this together so well done - looks a very nice job.
I've added a few pretty trivial comments but these points aside all seems to work well
The second test script on this PR should have a call to `remove_peaks_on_detector_edge` instead of `filter_peaks_on_detector_edge` so worth updating that in case people use this PR as a reference

RichardWaiteSTFC committed:
Review comments

- Add copyright headers
- Add doc string to sxd_calibrate_panels
- Delete unused/debug workspaces from the ADS
- Remove grp_ws and ngrps attriibutes from WishSX (unused) and SXD (now passed as func args)
- ConvertUnits back to TOF after absorption correction
- Add cropping limits as variable at top of func
- Add separate getters for wsname and ws object

RichardWaiteSTFC committed:
Update test data (now processed runs always output in TOF)

RichardWaiteSTFC committed:
Bug fixes: vanadium units for WISH and checking number gonio angles

RichardWaiteSTFC committed:
Add required memory in WISH test and increase tol in SXD test

RichardWaiteSTFC committed:
For WISH use subset of vanadium spectra to use less RAM in tests

To do this added ext argument to load_data so could load .nxs file with subset of spectra

RichardWaiteSTFC committed:
Fix bug when loading isaw UB file

RichardWaiteSTFC committed:
Update vaandium reference for WISH

RichardWaiteSTFC committed:
Expose SCDCalibratePanel kwrags to sxd panel calibration method

Also added function to reset/undo calibration (LoadInstrument does not reset calibration)

RichardWaiteSTFC committed:
Specify IDF used in LoadEmptyInstrument and exposed to user

sf1919 commented:
@RichardWaiteSTFC  As there are what appears to be valid test failures we won't re-run the ones that are currently 'scheduled' as they will rerun when you push a new commit

RichardWaiteSTFC committed:
Fix bug predicting fractional peaks and add test

RichardWaiteSTFC committed:
Remove deleted argument in call to SXD calibration in system test

thomashampson commented:
pre-commit.ci run

jhaigh0 reviewed (approved):
Can approve now the tests have been fixed
